### PR TITLE
Fix list declarations bug

### PR DESCRIPTION
**Pull request by BorjaRequena**
We have noticed a bug where the content of declarations that did not have a whitespace immediately after their name was not properly parsed. For example, 
```lean
theorem example_ok (n : ℕ) : ... -- correctly extracted

theorem example_not_ok(n : ℕ) : ... -- the content starts at ": ℕ" not at "(n". 
```

I have changed the function to list all the declarations within a piece of Lean code that assumed a whitespace after the name to actually split the declaration by whitespace and any special character that is not a letter, number or _. Then, I have overhauled the tests for these functions and updated them to include an example of this edge case. It should be fine now 😄

### CODE REVIEW

File: src/ax_prover/utils/lean_parsing.py [hunk 1]
```diff
@@ -326,7 +326,12 @@ def list_all_declarations_in_lean_code(raw_code: str) -> list[Declaration]:
     code = strip_comments(raw_code)
 
     for line in code.split("\n"):
-        line_keywords = line.strip().split()
+        # Split on any whitespace or any non-word, non-number character (preserve underscores, numbers, and unicode letters)
```
Comment (austinletson) [inline src/ax_prover/utils/lean_parsing.py]: As we discussed on slack there are a few cases that this doesn't cover. For example unicode chars or dots in identifiers (`def αβ_γ := 42` or `theorem Poly.not_principal`

What if instead of tokenizing the line we use a single regex to match the whole line then use capture groups to extract what we want. In other words focus on what is allowed in the identifier as opposed to what is allowed surrounding the identifier. We can use the full spec that is in the Language reference to build up the capture group for the identifier: https://lean-lang.org/doc/reference/latest/Source-Files-and-Modules/#keywords-and-identifiers

For example:
```python
decl_re = re.compile(r"(\w+)\s+([\w.?'!]+)\s*(.*)")
                                                                                                                                                                                                                      
for line in code.split("\n"):
    m = decl_re.match(line.strip())                                                                                                                                                                                   
    if m and m.group(1) in list(DeclarationType):                                                                                                                                                                   
        if declaration is not None:
            declarations.append(declaration)
        declaration = Declaration(
            declaration_type=m.group(1),
            name=m.group(2),
            content=m.group(3),
        )
```

Comment (austinletson) [reply src/ax_prover/utils/lean_parsing.py]: Also, should we add tests for some more of these cases: unicode chars, dots, new lines (I think my above code will break with new lines in the declaration)?

Comment (BorjaRequena) [reply src/ax_prover/utils/lean_parsing.py]: Yeah I think this is a much more robust approach. When I was developing the solution, I was thinking this would actually be better.

Comment (BorjaRequena) [reply src/ax_prover/utils/lean_parsing.py]: I have changed the implementation and I have expanded the tests to include more cases. However, I am not sure whether the tests are comprehensive enough. Please, feel free to suggest additional potential cases, currently, we cover `names_with_underscores`, `names.with.dots`, having spaces or not between name and subsequent special characters like ( or :, and having names with ' or ? in the name.

I believe the current pattern handles (hopefully) most cases except for the double guillemets '« »' specifications. I wonder how common they may be, I have never seen them. What do you think @austinletson?

Comment (austinletson-ax) [reply src/ax_prover/utils/lean_parsing.py]: Two comments here:

1. One more test suggestion is a name with random unicode chars, e.g., `def αβ_γ := 42`, but I would have a friend generate one with a bunch more that are allowed in Lean.

2. The double guillements is unfortunately a possibility, especially with older Lean projects where this used to be automatically generated more often. I think it is safest to include this even though it is an edge case. I think this is supported with the following updated regex: `decl_re = re.compile(r"(\w+)\s+((?:«[^»]+»|[\w.?'!])+)\s*(.*)")`, the idea here being to add an alternation to the capture group to match an id in guillements. Maybe one or two tests cases with the guillements and we should be good here. Thanks for the thoroughness!

**Comment by BorjaRequena** [comment]
I have added a few additional test cases covering guillemets and unicode characters of different groups. I found that mathematical symbols are not captured as `\w` in python, so I felt it was easier to try to catch everything between delimiters (`:`, `(`, `[`,...) than trying to enumerate everything we could get. This still only captures declarations of the form `<type> <name> <content>`.

Given that we would need to add plenty of special cases like non-computable objects, nameless declarations, and declarations that break this type-name-content pattern, such as, `instance {α : Type} [Add α] : Add (Point α) where add p q := { x := p.x + q.x, y := p.y + q.y }`, I believe it's better that we move on to extracting the declarations using Lean itself. The regex approach was nice when testing on curated benchmarks where we knew we'd only have a few definitions, theorems and lemmas, but may become a problem when dealing wiith projects on the wild.

**Review by austinletson — APPROVED** [review]
Looks good to me.

This will solve the current bugs, and in the future we can reimplement using LeanInteract for a more robust approach.
